staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsv1beta1client "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1"
+)
+
+var (
+	// retries is the maximum number of attempts that a CRD will be given to establish
+	retries = 5
+	// interval is the amount of time that will pass between retries
+	interval = 500 * time.Millisecond
+	// factor is the scalar by which the prevous interval will be increased
+	factor = 1.0
+)
+
+// EnsureCRDs is a utility function that will ensure that a list of CustomResourceDefinitions
+// are not only created, but ready for use by a Kubernetes cluster. If a CustomResourceDefinition
+// cannot be established within a reasonable amount of retries, an error will be returned.
+func EnsureCRDs(
```

Comment:
the name does not say what it ensures. I would also expect to have a stopCh to cancel it.

Reply:
We traditionally don't move helpers into the clients. I tend to leave it separate.

Another point: why do we need the slice of CRDs in that func? IMO the func should be for one CRD and the user can do any concurrency magic around.

Reply:
There are some benefits with adding it to the expansion interface:
* Can be versioned with the rest of the API. By keeping it separate in a utils pkg, you can only have one version at a time which is not scalable for future versions of the CRD API.
* It directly depends on the client and flows a lot better by having it as a receiver of the client.
* Fits nicely as a resource verb (ensure) in that API.

If tradition is the only reason why we're hesitant about it, I'd question what the definition would be for a method that would be a good fit for the expansion API.

Concerning the slice of CRDs, I guess I don't object to just operating on a single one and letting the client orchestrate multiple. They'd only have to add a few lines of code anyways.

Reply:
> Can be versioned with the rest of the API. By keeping it separate in a utils pkg, you can only have one version at a time which is not scalable for future versions of the CRD API.

It can still be a helper inside of the package. Just not a method on the client.

> It directly depends on the client and flows a lot better by having it as a receiver of the client.

Not a big difference to call `client.Ensure` or `clienthelper.Ensure(client, ...)`.

> Fits nicely as a resource verb (ensure) in that API.

I don't think so. Client methods correspond to API verbs and subresources. Ensure is neither an API verb, not a subresource. The expansion interface is meant to allow custom verb implementations, e.g. like podclient.Logs or podclient.Exec.